Compile with MinGW-64 (G++)

Hi fpagliughi,

Would you happen to know how to compile a sockpp included CPP file properly with G++? I've tried many including: 
g++ main.cpp -o main -I "sockpp/include/" -L "sockpp/build/Debug/sockpp-static.lib"

Including both the path and the linker file location. The program compiles properly when I have the actual include statement (example: #include <sockpp/tcp_connector.h>) only - but when trying to add a TCP connector (example: sockpp::tcp_connector conn({"localhost", 12345});) - it fails to compile.

So I tried running this compile command:
g++ main.cpp -o main -I "sockpp/include/" -L "sockpp/build/Debug/sockpp-static.lib" -lsockpp

But it was unable to identify what "-lsockpp" is, outputting: cannot find -lsockpp (returned 1 exit status) on Windows.

It is definitely a bit trickier on windows to perform these types of tasks but would you know how I could get MinGW to recognize what -lsockpp is?

--- a/.github/workflows/cmake-multi-platform.yml
+++ b/.github/workflows/cmake-multi-platform.yml
@@ -251,3 +251,45 @@ jobs:
         ASAN_OPTIONS: strict_string_checks=1:detect_stack_use_after_return=1:check_initialization_order=1:strict_init_order=1
         UBSAN_OPTIONS: print_stacktrace=1:halt_on_error=1
       run: ctest --output-on-failure
+
+  mingw:
+    runs-on: windows-latest
+
+    strategy:
+      fail-fast: false
+      matrix:
+        build_type: [Release, Debug]
+
+    defaults:
+      run:
+        shell: msys2 {0}
+
+    steps:
+    - uses: actions/checkout@v4
+
+    - name: Setup MSYS2
+      uses: msys2/setup-msys2@v2
+      with:
+        msystem: MINGW64
+        update: true
+        install: >-
+          mingw-w64-x86_64-gcc
+          mingw-w64-x86_64-cmake
+          mingw-w64-x86_64-catch2
+          mingw-w64-x86_64-ninja
+
+    - name: Configure CMake
+      run: >
+        cmake -B build -G Ninja
+        -DCMAKE_BUILD_TYPE=${{ matrix.build_type }}
+        -DSOCKPP_BUILD_SHARED=OFF
+        -DSOCKPP_BUILD_STATIC=ON
+        -DSOCKPP_BUILD_EXAMPLES=ON
+        -DSOCKPP_BUILD_TESTS=ON
+
+    - name: Build
+      run: cmake --build build
+
+    - name: Test
+      working-directory: build
+      run: ctest --output-on-failure

--- a/.github/workflows/cmake-multi-platform.yml
+++ b/.github/workflows/cmake-multi-platform.yml
@@ -275,7 +275,7 @@ jobs:
         install: >-
           mingw-w64-x86_64-gcc
           mingw-w64-x86_64-cmake
-          mingw-w64-x86_64-catch2
+          mingw-w64-x86_64-catch
           mingw-w64-x86_64-ninja
 
     - name: Configure CMake

--- a/include/sockpp/platform.h
+++ b/include/sockpp/platform.h
@@ -53,6 +53,10 @@
     // #pragma warning(4 : 4996)	// Deprecated functions (CRT & all)
     // #pragma warning(4 : 4250)	// Inheritance via dominance
 
+    #if !defined(_WIN32_WINNT)
+        #define _WIN32_WINNT 0x0600
+    #endif
+
     #if !defined(WIN32_LEAN_AND_MEAN)
         #define WIN32_LEAN_AND_MEAN
     #endif

--- a/include/sockpp/result.h
+++ b/include/sockpp/result.h
@@ -197,10 +197,44 @@ public:
      * Retrieves the last error from an operation.
      * This should be called after a failed system call to get the cause of
      * the error.
+     *
+     * On Windows, Winsock error codes are translated to their std::errc
+     * equivalents (generic_category) so that comparisons against std::errc
+     * values work correctly under both MSVC and MinGW-w64. MinGW's
+     * system_category is POSIX-based and does not map Winsock codes, so
+     * leaving them under system_category would cause those comparisons to
+     * silently fail throughout the library.
      */
     static error_code last_error() {
         int err = get_last_errno();
+#if defined(_WIN32)
+        switch (err) {
+            case WSAEWOULDBLOCK:   return std::make_error_code(errc::operation_would_block);
+            case WSAEINPROGRESS:   return std::make_error_code(errc::operation_in_progress);
+            case WSAEALREADY:      return std::make_error_code(errc::already_connected);
+            case WSAENOTSOCK:      return std::make_error_code(errc::not_a_socket);
+            case WSAECONNREFUSED:  return std::make_error_code(errc::connection_refused);
+            case WSAECONNRESET:    return std::make_error_code(errc::connection_reset);
+            case WSAECONNABORTED:  return std::make_error_code(errc::connection_aborted);
+            case WSAETIMEDOUT:     return std::make_error_code(errc::timed_out);
+            case WSAEHOSTUNREACH:  return std::make_error_code(errc::host_unreachable);
+            case WSAENETUNREACH:   return std::make_error_code(errc::network_unreachable);
+            case WSAEADDRINUSE:    return std::make_error_code(errc::address_in_use);
+            case WSAEADDRNOTAVAIL: return std::make_error_code(errc::address_not_available);
+            case WSAEMSGSIZE:      return std::make_error_code(errc::message_size);
+            case WSAEACCES:        return std::make_error_code(errc::permission_denied);
+            case WSAEFAULT:        return std::make_error_code(errc::bad_address);
+            case WSAEINVAL:        return std::make_error_code(errc::invalid_argument);
+            case WSAEMFILE:        return std::make_error_code(errc::too_many_files_open);
+            case WSAEISCONN:       return std::make_error_code(errc::already_connected);
+            case WSAENOTCONN:      return std::make_error_code(errc::not_connected);
+            case WSAESHUTDOWN:     return std::make_error_code(errc::broken_pipe);
+            case WSAENETDOWN:      return std::make_error_code(errc::network_down);
+            default:               return error_code{err, std::system_category()};
+        }
+#else
         return error_code{err, std::system_category()};
+#endif
     }
     /**
      * Determines if the result represents a failed operation.

--- a/src/inet_address.cpp
+++ b/src/inet_address.cpp
@@ -108,11 +108,9 @@ inet_address::inet_address(const sock_address& addr, error_code& ec) noexcept {
 // --------------------------------------------------------------------------
 
 result<in_addr_t> inet_address::resolve_name(const string& saddr) noexcept {
-#if !defined(_WIN32)
     in_addr ia;
     if (::inet_pton(ADDRESS_FAMILY, saddr.c_str(), &ia) == 1)
         return ia.s_addr;
-#endif
 
     addrinfo *res, hints = addrinfo{};
     hints.ai_family = ADDRESS_FAMILY;
@@ -123,7 +121,7 @@ result<in_addr_t> inet_address::resolve_name(const string& saddr) noexcept {
     if (err != 0) {
         error_code ec{};
 #if defined(_WIN32)
-        ec = error_code{errno, system_category()};
+        ec = error_code{err, system_category()};
 #else
         if (err == EAI_SYSTEM)
             ec = result<>::last_error();
@@ -235,11 +233,9 @@ inet6_address::inet6_address(const sock_address& addr, error_code& ec) noexcept 
 // --------------------------------------------------------------------------
 
 result<in6_addr> inet6_address::resolve_name(const string& saddr) noexcept {
-#if !defined(_WIN32)
     in6_addr ia;
     if (::inet_pton(ADDRESS_FAMILY, saddr.c_str(), &ia) == 1)
         return ia;
-#endif
 
     addrinfo *res, hints = addrinfo{};
     hints.ai_family = ADDRESS_FAMILY;

